interface/ceed-basis.c

```diff
@@ -1085,6 +1083,110 @@ int CeedBasisApply(CeedBasis basis, CeedInt num_elem, CeedTransposeMode t_mode,
+                           CeedVector v) {
+  CeedSize x_length = 0, u_length = 0, v_length;
+  CeedInt  dim, num_comp, num_nodes;
+  CeedCall(CeedBasisGetDimension(basis, &dim));
+  CeedCall(CeedBasisGetNumComponents(basis, &num_comp));
+  CeedCall(CeedBasisGetNumNodes(basis, &num_nodes));
+  CeedCall(CeedVectorGetLength(x_ref, &x_length));
+  CeedCall(CeedVectorGetLength(v, &v_length));
+  if (u) CeedCall(CeedVectorGetLength(u, &u_length));
+
+  // Check compatibility of topological and geometrical dimensions
+  if ((t_mode == CEED_TRANSPOSE && (v_length % num_nodes != 0 || u_length != (num_points * num_comp))) ||
+      (t_mode == CEED_NOTRANSPOSE && (u_length % num_nodes != 0 || v_length != (num_points * num_comp)))) {
+    // LCOV_EXCL_START
+    return CeedError(basis->ceed, CEED_ERROR_DIMENSION, "Length of input/output vectors incompatible with basis dimensions and number of points");
+    // LCOV_EXCL_STOP
+  }
+
+  // Check compatibility coordinates vector
+  if (x_length != num_points * dim) {
+    // LCOV_EXCL_START
+    return CeedError(basis->ceed, CEED_ERROR_DIMENSION,
+                     "Length of reference coordinate vector incompatible with basis dimension and number of points");
+    // LCOV_EXCL_STOP
+  }
+
+  // Check vector lengths to prevent out of bounds issues
+  bool bad_dims = false;
+  switch (eval_mode) {
+    case CEED_EVAL_INTERP:
+      bad_dims = ((t_mode == CEED_TRANSPOSE && (u_length < num_points * num_comp || v_length < num_comp * num_nodes)) ||
+                  (t_mode == CEED_NOTRANSPOSE && (v_length < num_points * num_comp || u_length < num_comp * num_nodes)));
+      break;
+    case CEED_EVAL_GRAD:
+      bad_dims = ((t_mode == CEED_TRANSPOSE && (u_length < num_points * num_comp * dim || v_length < num_comp * num_nodes)) ||
+                  (t_mode == CEED_NOTRANSPOSE && (v_length < num_points * num_comp * dim || u_length < num_comp * num_nodes)));
+      break;
+    case CEED_EVAL_NONE:
+    case CEED_EVAL_WEIGHT:
+    case CEED_EVAL_DIV:
+    case CEED_EVAL_CURL:
+      // LCOV_EXCL_START
+      return CeedError(basis->ceed, CEED_ERROR_UNSUPPORTED, "Evaluation at arbitrary points not supported for %s", CeedEvalModes[eval_mode]);
+      // LCOV_EXCL_STOP
+  }
+  if (bad_dims) {
+    // LCOV_EXCL_START
+    return CeedError(basis->ceed, CEED_ERROR_DIMENSION, "Input/output vectors too short for basis and evaluation mode");
+    // LCOV_EXCL_STOP
+  }
+
+  // Backend method
+  if (basis->ApplyAtPoints) {
+    CeedCall(basis->ApplyAtPoints(basis, num_points, t_mode, eval_mode, x_ref, u, v));
+    return CEED_ERROR_SUCCESS;
+  }
+
+  // Default implementation
+  if (!basis->legendre_coeffs_1d) {
+    // Build matrix mapping from quadrature point values to Legendre coefficients
+  }
+  if (t_mode == CEED_NOTRANSPOSE) {
+    // Nodes to arbitrary points
+    CeedVector q;
+
+    // -- Interpolate to quadrature points
+    CeedCall(CeedVectorCreate(basis->ceed, basis->Q * basis->Q_comp, &q));
```

Comment:
Probably will want to cache this vector - GPU backends will do this work on the device